core/src/main/java/tech/tablesaw/conversion/smile/SmileConverter.java

```diff
@@ -22,125 +24,62 @@ public SmileConverter(Relation table) {
-        table.columns(variablesColNames));
-  }
-
-  private AttributeDataset dataset(
-      NumericColumn<?> responseCol, AttributeType type, List<Column<?>> variableCols) {
-    List<Column<?>> convertedVariableCols =
-        variableCols.stream()
-            .map(col -> col.type() == ColumnType.STRING ? col : table.nCol(col.name()))
-            .collect(Collectors.toList());
-    Attribute responseAttribute =
-        type == AttributeType.NOMINAL
-            ? colAsNominalAttribute(responseCol)
-            : new NumericAttribute(responseCol.name());
-    AttributeDataset dataset =
-        new AttributeDataset(
-            table.name(),
-            convertedVariableCols.stream()
-                .map(col -> colAsAttribute(col))
-                .toArray(Attribute[]::new),
-            responseAttribute);
-    for (int i = 0; i < responseCol.size(); i++) {
-      final int r = i;
-      double[] x =
-          IntStream.range(0, convertedVariableCols.size())
-              .mapToDouble(c -> getDouble(convertedVariableCols.get(c), dataset.attributes()[c], r))
-              .toArray();
-      dataset.add(x, responseCol.getDouble(r));
-    }
-    return dataset;
-  }
-
-  private double getDouble(Column<?> col, Attribute attr, int r) {
-    if (col.type() == ColumnType.STRING) {
-      String value = ((StringColumn) col).get(r);
-      try {
-        return attr.valueOf(value);
-      } catch (ParseException e) {
-        throw new IllegalArgumentException("Error converting " + value + " to nominal", e);
+  public DataFrame toDataFrame(StructType schema) {
+    List<Tuple> rows = new ArrayList<>();
+    int colCount = table.columnCount();
+    for (int rowIndex = 0; rowIndex < table.rowCount(); rowIndex++) {
+      Object[] row = new Object[colCount];
+      for (int colIndex = 0; colIndex < colCount; colIndex++) {
+        Column<?> col = table.column(colIndex);
+        if (!col.isMissing(rowIndex)) {
+          row[colIndex] =
+              col.type().equals(ColumnType.INSTANT)
+                  ? LocalDateTime.ofInstant(((InstantColumn) col).get(rowIndex), ZoneOffset.UTC)
+                  : col.get(rowIndex);
+        }
       }
+      rows.add(Tuple.of(row, schema));
     }
-    if (col instanceof NumericColumn) {
-      return ((NumericColumn<?>) col).getDouble(r);
-    }
-    throw new IllegalStateException(
-        "Error converting " + col.type() + " column " + col.name() + " to Smile");
-  }
-
-  private Attribute colAsAttribute(Column<?> col) {
-    return col.type() == ColumnType.STRING
-        ? colAsNominalAttribute(col)
-        : new NumericAttribute(col.name());
-  }
 
-  private NominalAttribute colAsNominalAttribute(Column<?> col) {
-    Column<?> unique = col.unique();
-    return new NominalAttribute(
-        col.name(),
-        unique
-            .mapInto(o -> o.toString(), StringColumn.create(col.name(), unique.size()))
-            .asObjectArray());
+    schema = schema.boxed(rows);
```

Comment:
@haifengl I was wondering if this made sense. I copied it from `CSV.java`, but should we not use the `schema` that's already available? I was curious why `CSV.read` does this instead of using the user-provided schema. Won't we lose column names, etc.?

Reply:
It is in case that missing values exist but we didn't detect it during inferSchema. For user provided schema, it should be nop (if the user provides the correct schema).